src/NationalInstruments.Analyzers/Correctness/ApprovedNamespaceAnalyzer.cs

```diff
@@ -0,0 +1,268 @@
+        /// </summary>
+        public static DiagnosticDescriptor FileReadRule { get; } = new DiagnosticDescriptor(
+            $"{DiagnosticId}_ReadError",
+            new LocalizableResourceString(nameof(Resources.NI1800_FileReadErrorTitle), Resources.ResourceManager, typeof(Resources)),
+            new LocalizableResourceString(nameof(Resources.NI1800_FileReadErrorMessage), Resources.ResourceManager, typeof(Resources)),
+            Resources.CategoryNamespaces,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <summary>
+        /// Rule for missing approved namespaces files.
+        /// </summary>
+        public static DiagnosticDescriptor MissingApprovalFilesRule { get; } = new DiagnosticDescriptor(
+            $"{DiagnosticId}_ReadError",
+            new LocalizableResourceString(nameof(Resources.NI1800_MissingApprovalFilesErrorTitle), Resources.ResourceManager, typeof(Resources)),
+            new LocalizableResourceString(nameof(Resources.NI1800_MissingApprovalFilesErrorMessage), Resources.ResourceManager, typeof(Resources)),
+            Resources.CategoryNamespaces,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <inheritdoc/>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(ProductionRule, TestRule, FileReadRule, MissingApprovalFilesRule);
+
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecutionIf(IsRunningInProduction && !InDebugMode);
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        /// <summary>
+        /// Adds a namespace to the approved namespaces list
+        /// </summary>
+        /// <param name="namespaceName">Namespace to be approved</param>
+        internal static void ApproveNamespace(string namespaceName)
+        {
+            var namespacesFilePath = IsTestNamespace(namespaceName) ? _approvedTestNamespacesFilePath : _approvedNamespacesFilePath;
+            var lines = File.ReadAllLines(namespacesFilePath);
+            var namespaces = lines
+                    .Concat(new[] { namespaceName })
+                    .Select(x => x.Trim())
+                    .OrderBy(x => x)
+                    .Distinct();
+            File.WriteAllLines(namespacesFilePath, namespaces);
+        }
+
+        private static bool IsTestNamespace(string namespaceName)
+        {
+            return _testNamespacePattern.IsMatch(namespaceName);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationStartContext)
+        {
+            InitializeApprovedNamespaces();
+            compilationStartContext.RegisterSymbolAction(AnalyzeNamespace, SymbolKind.Namespace);
+
+            if (!ApprovalFilesExist())
```

Comment:
Thanks for making this change, though it'd probably be more helpful to report diagnostics in `else` cases on lines 154/166. Then you can be more granular and explain that a file named "ApprovedNamespaces.txt" or "ApprovedNamespaces.Tests.txt" is missing. But, that's up to you.

Reply:
Those files will change going forward. I will leave it as it is for now and iterate on it as new changes are applied